### PR TITLE
C++20 compatibility

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

This doesn't yet bump the version requirement, but makes all the changes currently necessary to compile with C++20. We should do some simulations to make sure we don't have performance regressions, etc. before actually bumping.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Parallel/GlobalCache.hpp [hunk 1]
```diff
@@ -119,6 +119,23 @@ struct MockGlobalCache {
...
   virtual ~MockGlobalCache() = default;
   virtual void pup(PUP::er& /*p*/) {}
+
+  static bool isIrreducible() { return true; }
+
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wsuggest-attribute=noreturn"
+#elif defined(__clang__)
+#pragma GCC diagnostic push
```
Comment (nilsvu) [inline src/Parallel/GlobalCache.hpp:129]: Is this missing a `pop`?

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/Domain/Structure/ElementId.cpp [hunk 2]
```diff
@@ -25,9 +25,15 @@
 // - `ElementId` must satisfy `std::is_pod`
 // - `ElementId` must not be larger than the size of three `int`s, i.e.
 //   `sizeof(ElementId) <= 3 * sizeof(int)`
-static_assert(std::is_pod<ElementId<1>>::value, "ElementId is not POD");
-static_assert(std::is_pod<ElementId<2>>::value, "ElementId is not POD");
-static_assert(std::is_pod<ElementId<3>>::value, "ElementId is not POD");
+static_assert(std::is_standard_layout_v<ElementId<1>> and
+                  std::is_trivial_v<ElementId<1>>,
```
Comment (wthrowe) [inline src/Domain/Structure/ElementId.cpp]: Update comment.  (Still refers to `is_pod`.)

File: src/Parallel/GlobalCache.hpp [hunk 3]
```diff
@@ -119,6 +119,23 @@ struct MockGlobalCache {
...
+#pragma GCC diagnostic ignored "-Wmissing-noreturn"
+#endif  // defined(__GNUC__) && !defined(__clang__)
+  template <typename T>
+  static void contribute(T&& /*unused*/) {
```
Comment (wthrowe) [inline src/Parallel/GlobalCache.hpp:133]: Why not mark it as noreturn instead of silencing the warning?

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Parallel/GlobalCache.hpp (hunk 3 above)
Comment (nilsdeppe) [reply src/Parallel/GlobalCache.hpp:133]: Charm++ compatibility. I don't want to worry about having anything different in the interface than what Charm++ does.

**Comment by nilsdeppe** [comment]
Pushed a fixup. Thanks for the reviews!

**Comment by wthrowe** [comment]
Looks good.

**Comment by nilsdeppe** [comment]
@nilsvu could you take a look at the fixup when you have a minute? I'd like to start poking at enabling C++20 in CI maybe over the weekend or next week.

**Review by nilsvu — APPROVED** [review]

**Comment by nilsdeppe** [comment]
Squashed and rebased on develop. Thanks for the reviews :)

**Review by wthrowe — APPROVED** [review]

**Pull request merged**